core/services/relay/evm/binding.go

```diff
@@ -0,0 +1,156 @@
+package evm
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/common"
+
+	commontypes "github.com/smartcontractkit/chainlink-common/pkg/types"
+
+	evmclient "github.com/smartcontractkit/chainlink/v2/core/chains/evm/client"
+	"github.com/smartcontractkit/chainlink/v2/core/chains/evm/logpoller"
+	evmtypes "github.com/smartcontractkit/chainlink/v2/core/chains/evm/types"
+)
+
+type readBinding interface {
+	GetLatestValue(ctx context.Context, params any) ([]byte, error)
+	Bind(binding commontypes.BoundContract) error
+	SetCodec(codec commontypes.Codec)
+	Register() error
+	Unregister() error
+}
+
+type methodBinding struct {
+	address      common.Address
+	contractName string
+	method       string
+	client       evmclient.Client
+	codec        commontypes.Codec
+	bound        bool
+}
+
+var _ readBinding = &methodBinding{}
+
+func (m *methodBinding) SetCodec(codec commontypes.Codec) {
+	m.codec = codec
+}
+
+func (m *methodBinding) Register() error {
+	return nil
+}
+
+func (m *methodBinding) Unregister() error {
+	return nil
+}
+
+func (m *methodBinding) GetLatestValue(ctx context.Context, params any) ([]byte, error) {
+	if !m.bound {
+		return nil, fmt.Errorf("%w: method not bound", commontypes.ErrInvalidType)
+	}
+
+	data, err := m.codec.Encode(ctx, params, wrapItemType(m.contractName, m.method, true))
+	if err != nil {
+		return nil, err
+	}
+
+	callMsg := ethereum.CallMsg{
+		To:   &m.address,
+		From: m.address,
+		Data: data,
+	}
+
+	return m.client.CallContract(ctx, callMsg, nil)
```

Comment:
Originally, we'd envisioned having a caching param in the `RelayConfig`, where `caching=true` means it keeps polling the blockchain for the value and updating it in memory... returning the latest value when it's called.

We'll definitely need that on Solana, but not sure whether we'll ever need it on evm.  Either way, we can skip it for now on evm since it's not needed for median.

I'm wondering whether this binding interface will be flexible enough to handle something like that, if we do have to add it later.  I think so, but noting here in case anyone else has some input to add to that or a different perspective.

Reply:
I wonder if there should be an implementation of a client that does the caching.  Then it can be reused by anyone.  Eg something like this: 

```

type CachingChainReader struct {
   Underlying types.ChainReader
   Cache types.Cache // or just something that already exists
}

var _ types.ChainReader = &CachingChainReader{}

func (cr *CachingChainReader) GetLatestValue(ctx context.Context, contractName, method string, params any, returnVal any) error {
    value, ok := // lookup method + params in cache (could be many kinds of cache if it's an interface), return value if there (and still valid pending on the cache implementation)
    if ok {
      return value
     }
     value, err := cr.Underlying.GetLatestValue(ctx, contractName, method, params)
     if err != nil {
         // depending on implementation or config, a stale hit from the cache could be preferred to an error in this case?
         return err
     }
     // insert value to cache
     return value, nil
}
```

Reply:
Ah, re-read what you're describing and realized the continual polling wasn't in there.  I think that can still be wrapped with this interface.  The cache implementation would just have a `types.ChainReader` that it calls in a loop with all keys its seen before on a `go routine`.